misoc/targets/kasli.py

```diff
@@ -34,71 +36,114 @@ def __init__(self, cd, async_reset):
         ]
 
 
-class _CRG(Module):
-    def __init__(self, platform):
+class _CRG(Module, AutoCSR):
+    def __init__(self, platform, freq=125.0e6):
         self.clock_domains.cd_sys = ClockDomain()
         self.clock_domains.cd_sys4x = ClockDomain(reset_less=True)
         self.clock_domains.cd_sys4x_dqs = ClockDomain(reset_less=True)
         self.clock_domains.cd_clk200 = ClockDomain()
 
+        self.clock_sel = CSRStorage()
+        self.pll_locked = CSRStatus()
+        self.mmcm_locked = CSRStatus()
+
+        # old "sys" clock, now bootstrap
         clk125 = platform.request("clk125_gtp")
         platform.add_period_constraint(clk125, 8.)
         self.clk125_buf = Signal()
-        self.clk125_div2 = Signal()
+        clk125_div2 = Signal()
         self.specials += Instance("IBUFDS_GTE2",
             i_CEB=0,
             i_I=clk125.p, i_IB=clk125.n,
             o_O=self.clk125_buf,
-            o_ODIV2=self.clk125_div2)
+            o_ODIV2=clk125_div2)
+
+        # "rtio" clock (mgt flavor)
+        if platform.hw_rev == "v2.0":
+            cdr_clk_out = platform.request("cdr_clk_clean")
+        else:
+            cdr_clk_out = platform.request("si5324_clkout")
+
+        if freq == 125.0e6:
+           cdr_clk_period = 8.
+        elif freq == 100.0e6:
+            cdr_clk_period = 10.
+        else:
+            raise NotImplementedError
+         
+        self.cdr_clk_buf = Signal()
```

Comment:
Where is this attribute used?

Reply:
OK, I see, you are moving the IBUFDS_GTE2 from ARTIQ gtx_7series.py to here?

Reply:
Simpler to use cdr_clk_clean_fabric / si5324_clkout_fabric.
BTW this will break DRTIO since you won't meet timing at the transceiver's TXDATA pins anymore. On DRTIO systems, the post-bootstrap system clock needs to be driven from the transceiver's TXOUTCLK pin. But we can deal with that later.

Reply:
You will need this (from UG482).
![image](https://user-images.githubusercontent.com/720864/186361655-17cbd76d-67d8-4421-b5e8-52a3099da205.png)
BUFH if you can use it would be lighter on resources.